### PR TITLE
*: Truncate partition failpoint tests

### PR DESCRIPTION
**Pull request by mjonss**
### What problem does this PR solve?

Issue Number: ref #55819

Problem Summary:
Added failure injection tests.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/ddl/tests/partition/error_injection_test.go [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+	alter := `alter table t truncate partition p0,p2`
+	beforeDML := []string{
+		`insert into t values (1,1,1),(2,2,2),(4,4,4),(8,8,8),(9,9,9),(6,6,6)`,
+		`update t set a = 7, b = 7, c = 7 where a = 1`,
+		`update t set b = 3, c = 3 where c = 4`,
+		`delete from t where a = 8`,
+		`delete from t where b = 2`,
+	}
+	beforeResult := testkit.Rows("4 3 3", "6 6 6", "7 7 7", "9 9 9")
+	//beforeResult := testkit.Rows("4 3 3", "6 6 6", "7 7 7", "9 9 9")
```
Comment (Defined2014) [inline pkg/ddl/tests/partition/error_injection_test.go:78]: remove this line.

File: pkg/ddl/tests/partition/error_injection_test.go [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+	require.Equal(t, 0, len(pi.NewPartitionIDs), name)
+	require.Equal(t, len(tOrg.Meta().Indices), len(tt.Meta().Indices), name)
+	for i := range tOrg.Meta().Indices {
+		require.Equal(t, idxIDs[i], tt.Meta().Indices[i].ID, name)
+	}
+	for i, pid := range pids {
+		require.Equal(t, pid, tt.Meta().Partition.Definitions[i].ID, name)
+	}
+	tk.MustExec(`admin check table t /* ` + name + ` */`)
+	// TMP debug
```
Comment (Defined2014) [inline pkg/ddl/tests/partition/error_injection_test.go:200]: Need remove those lines or not?

File: pkg/ddl/rollingback.go [hunk 3]
```diff
@@ -316,6 +316,46 @@ func rollingbackExchangeTablePartition(jobCtx *jobContext, job *model.Job) (ver
 	return ver, errors.Trace(err)
 }
 
+func rollingbackTruncateTablePartition(jobCtx *jobContext, job *model.Job) (ver int64, err error) {
+	if job.SchemaState == model.StatePublic {
+		job.State = model.JobStateCancelled
+		return ver, dbterror.ErrCancelledDDLJob
+	}
```
Comment (Defined2014) [inline pkg/ddl/rollingback.go:323]: Could remove those lines. Let the cancel logic all in `convertTruncateTablePartitionJob2RollbackJob` function.

File: pkg/ddl/rollingback.go [hunk 4]
```diff
@@ -316,6 +316,46 @@ func rollingbackExchangeTablePartition(jobCtx *jobContext, job *model.Job) (ver
...
+	tblInfo, err := GetTableInfoAndCancelFaultJob(jobCtx.metaMut, job, job.SchemaID)
+	if err != nil {
+		return ver, errors.Trace(err)
+	}
+
+	return convertTruncateTablePartitionJob2RollbackJob(jobCtx, job, dbterror.ErrCancelledDDLJob, tblInfo)
+}
+
+func convertTruncateTablePartitionJob2RollbackJob(jobCtx *jobContext, job *model.Job, otherwiseErr error, tblInfo *model.TableInfo) (ver int64, err error) {
+	if job.SchemaState != model.StatePublic && job.SchemaState != model.StateWriteOnly {
```
Comment (Defined2014) [inline pkg/ddl/rollingback.go:334]: Maybe we could use `!job.IsRollbackable()` here.

**Review by tangenta — APPROVED** [review]

**Pull request merged**